src/arch/arm/CodeGen.zig

```diff
@@ -1246,61 +1249,67 @@ fn airSliceElemVal(self: *Self, inst: Air.Inst.Index) !void {
-            }
+                const tag: Mir.Inst.Tag = switch (elem_size) {
+                    1 => .ldrb,
+                    4 => .ldr,
+                    else => unreachable,
+                };
+                const shift: u5 = switch (elem_size) {
+                    1 => 0,
+                    4 => 2,
+                    else => unreachable,
+                };
 
-            break :result MCValue{ .register = dst_reg };
-        } else {
-            const dst_mcv = try self.allocRegOrMem(inst, false);
+                _ = try self.addInst(.{
+                    .tag = tag,
+                    .data = .{ .rr_offset = .{
+                        .rt = dst_reg,
+                        .rn = base_mcv.register,
+                        .offset = .{ .offset = Instruction.Offset.reg(index_reg, .{ .lsl = shift }) },
+                    } },
+                });
 
-            const addr_reg = try self.register_manager.allocReg(null);
-            self.register_manager.freezeRegs(&.{addr_reg});
-            defer self.register_manager.unfreezeRegs(&.{addr_reg});
+                break :result dst_mcv;
+            },
+            else => {
+                const dst_mcv = try self.allocRegOrMem(inst, true);
+
+                const offset_mcv = try self.genArmMulConstant(bin_op.rhs, @intCast(u32, elem_size));
+                assert(offset_mcv == .register); // result of multiplication should always be register
+                self.register_manager.freezeRegs(&.{offset_mcv.register});
+                defer self.register_manager.unfreezeRegs(&.{offset_mcv.register});
 
-            try self.genArmBinOpCode(addr_reg, base_mcv, offset_mcv, false, .add, .unsigned);
+                const addr_reg = try self.register_manager.allocReg(null);
+                self.register_manager.freezeRegs(&.{addr_reg});
+                defer self.register_manager.unfreezeRegs(&.{addr_reg});
 
-            // I know we will unfreeze these registers at the end of
-            // the scope of :result. However, at this point in time,
-            // neither the base register nor the offset register
-            // contains any valuable data anymore. In order to reduce
-            // register pressure, unfreeze them prematurely
-            self.register_manager.unfreezeRegs(&.{ base_mcv.register, offset_mcv.register });
+                try self.genArmBinOpCode(addr_reg, base_mcv, offset_mcv, false, .add, .unsigned);
 
-            try self.load(dst_mcv, .{ .register = addr_reg }, slice_ptr_field_type);
+                // I know we will unfreeze these registers at the end of
+                // the scope of :result. However, at this point in time,
+                // neither the base register nor the offset register
+                // contains any valuable data anymore. In order to reduce
+                // register pressure, unfreeze them prematurely
+                self.register_manager.unfreezeRegs(&.{ base_mcv.register, offset_mcv.register });
```

Comment:
Hmm, this doesn't feel right to me. We should perhaps figure out a way to do this automatically. Also, we only make a call to `load` subsequently, so I guess reducing pressure here won't have a measurable effect?

Reply:
Yes, this does in fact have a measurable impact: Without this, the register manager will quickly run out of usable registers (of which there are only 6, see code snippet below) in certain test cases which involve inline `memcpy`.

https://github.com/ziglang/zig/blob/95fbce2b958395a367a82ce33170edd93e686173/src/arch/arm/bits.zig#L189

The long term solution to this consists of

- calling extern `memcpy`
- use caller-preserved registers as well

But for now (and possibly also for the future for reducing register pressure), this tiny optimization really does help.

Reply:
Right, agreed. So in this case, I believe your intent will be better communicated if you remove the defers in favour of explicit management of unfreezing (in this block). Then, there is no need for the comment here. Rather, we could add a comment to this effect to the `RegisterManager`'s docs to better document different idioms and uses. Lemme know what you think about this proposal!

Reply:
Ok, will do!

Reply:
@joachimschmidt557 I'm gonna merge this as-is and you can follow up with our proposal in a subsequent PR. Hope that sounds good to you!